src/passes/GUFA.cpp

```diff
@@ -0,0 +1,308 @@
+      return false;
+    }
+
+    // Pops are structurally necessary in catch bodies, and removing a try could
+    // leave a pop without a proper parent.
+    return !curr->is<Pop>() && !curr->is<Try>();
+  }
+
+  // Whether we can remove something (but not its children) without changing
+  // observable behavior or breaking validation.
+  bool canRemove(Expression* curr) {
+    if (!canRemoveStructurally(curr)) {
+      return false;
+    }
+    return !EffectAnalyzer(getPassOptions(), *getModule(), curr)
+              .hasUnremovableSideEffects();
+  }
+
+  // Whether we can replace something (but not its children, we can keep them
+  // with drops) with an unreachable without changing observable behavior or
+  // breaking validation.
+  bool canReplaceWithUnreachable(Expression* curr) {
+    if (!canRemoveStructurally(curr)) {
+      return false;
+    }
+    EffectAnalyzer effects(getPassOptions(), *getModule(), curr);
+    // Ignore a trap, as the unreachable replacement would trap too.
+    effects.trap = false;
+    return !effects.hasUnremovableSideEffects();
+  }
+
+  void visitExpression(Expression* curr) {
+    // Skip things we can't improve in any way.
+    auto type = curr->type;
+    if (type == Type::unreachable || type == Type::none ||
+        Properties::isConstantExpression(curr)) {
+      return;
+    }
+
+    if (type.isTuple()) {
+      // TODO: tuple types.
+      return;
+    }
+
+    if (type.isRef() && (getTypeSystem() != TypeSystem::Nominal &&
+                         getTypeSystem() != TypeSystem::Isorecursive)) {
+      // Without type info we can't analyze subtypes, so we cannot infer
+      // anything about refs.
+      return;
+    }
+
+    // Ok, this is an interesting location that we might optimize. See what the
+    // oracle says is possible there.
+    auto contents = oracle.getContents(ExpressionLocation{curr, 0});
+
+    auto& options = getPassOptions();
+    auto& wasm = *getModule();
+    Builder builder(wasm);
+
+    auto replaceWithUnreachable = [&]() {
+      if (canReplaceWithUnreachable(curr)) {
+        replaceCurrent(getDroppedChildrenAndAppend(
+          curr, wasm, options, builder.makeUnreachable()));
+      } else {
+        // We can't remove this, but we can at least put an unreachable
+        // right after it.
+        replaceCurrent(builder.makeSequence(builder.makeDrop(curr),
+                                            builder.makeUnreachable()));
+      }
+      optimized = true;
+    };
+
+    if (contents.getType() == Type::unreachable) {
+      // This cannot contain any possible value at all. It must be unreachable
+      // code.
+      replaceWithUnreachable();
+      return;
+    }
+
+    // This is reachable. Check if we can emit something optimized for it.
+    // TODO: can we handle more general things here too?
+    if (!contents.canMakeExpression()) {
+      return;
+    }
+
+    if (contents.isNull() && curr->type.isNullable()) {
```

Comment:
Can this ever be non-nullable when the contents is null?

Reply:
It should trap in that case before we get here (see comment on line 169 below, which should maybe be moved up here).

That is, a null can be cast to have a non-nullable type, and so the flow might think that a null can arrive here. On line 182 there is a TODO to filter out such values in the flow.